<a href="https://colab.research.google.com/github/nunososorio/bhs/blob/main/NSO_PracticalClass_II.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practical Training: Omics, Bioinformatics, and Pharmacogenomics

By **Nuno S. Osório** 🖋️

👋 Welcome to this tutorial! We will explore how to access and use omics-related databases. We will focus on retrieving data from **BioMart** using their respective Python packages, and then analyze the retrieved data. This tutorial is designed to be run on jupyter notebook environments and includes exercises that involve running Python code. 🐍💻

You can access an interactive cloud version of the notebook here (https://colab.research.google.com/github/nunososorio/bhs/blob/main/NSO_PracticalClass_II.ipynb).

Let's dive in! 🏊‍♂️


# Introduction to Omics and Its Role in Drug Development 🧪💊

Welcome to this practical exploration of omics and its significant role in drug development. 🎓🔬

## Omics 🧬

Omics is an integrative field of study in biology that encompasses disciplines such as genomics, proteomics, and metabolomics. The goal of omics is to collectively characterize and quantify pools of biological molecules, providing a holistic view of the structure, function, and dynamics of an organism. 🌐

## Omics in Drug Development 💡

In drug development, omics technologies play a crucial role in understanding the molecular mechanisms of diseases. This molecular-level understanding aids in the identification of potential drug targets and the development of effective therapeutic agents. 💊

## Databases in Omics Research 🗃️

Bioinformatics databases are essential resources in omics research. These databases, such as the [Ensembl](https://www.ensembl.org/index.html), and [UniProt](https://www.uniprot.org/), store a wealth of omics data. Accessing and analyzing this data is a critical step in drug discovery and development. 📚

## Programmatic Access to Databases 💻

For large-scale and reproducible analysis, programmatic access to these databases is often more efficient than manual data retrieval through web interfaces. Python, a popular language in bioinformatics, offers several libraries for this purpose. For instance, the pybiomart library provides an interface to the Ensembl, enabling the retrieval and analysis of omics data directly within Python scripts. 🐍

In the following sections of this tutorial, we will delve deeper into how to use this Python library for accessing and analyzing omics data. 🚀


## Setup
First, we need to install the necessary Python packages. Run the following commands in your environment:


In [ ]:
!pip install pybiomart


Now, import the library:

In [ ]:
import pybiomart as


## Explore

Lets start by learning all types of data or information we can retreive from Ensembl database using the pybiomart library:

In [16]:
from pybiomart import Server

# Connect to the server
server = Server(host='http://www.ensembl.org')

# Select the human dataset
dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['hsapiens_gene_ensembl']

# List all available attributes
attributes = dataset.list_attributes()
attributes



name                  display_name  \
0                   ensembl_gene_id                Gene stable ID   
1           ensembl_gene_id_version        Gene stable ID version   
2             ensembl_transcript_id          Transcript stable ID   
3     ensembl_transcript_id_version  Transcript stable ID version   
4                ensembl_peptide_id             Protein stable ID   
...                             ...                           ...   
3200    sequence_genomic_coding_end            Genomic coding end   
3201          sequence_constitutive                                 
3202                     cds_length                    CDS Length   
3203                      cds_start                     CDS start   
3204                        cds_end                       CDS end   

                                 description  
0                      Stable ID of the Gene  
1           Versionned stable ID of the Gene  
2                Stable ID of the Transcript  
3     Versionned stable ID of the Transcript  
4                                             
...                                      ...  
3200                                          
3201                                          
3202                                          
3203                                          
3204                                          

[3205 rows x 3 columns]

As you can see, it returns a table containing a wealth of information about different attributes (3205) of the gene.

Here’s a brief explanation of some of the attributes:

- **'ensembl_gene_id'**: The Ensembl ID of the gene.
- **'ensembl_transcript_id'**: The Ensembl ID of the transcript.
- **'description'**: The description of the gene.
- **'chromosome_name'**: The name of the chromosome where the gene is located.
- **'start_position'**: The start position of the gene on the chromosome.
- **'end_position'**: The end position of the gene on the chromosome.
- **'strand'**: The strand of the gene on the chromosome.
- **'transcript_count'**: The number of transcripts of the gene.
- **'percentage_gc_content'**: The percentage of GC content in the gene.



We can query the database selecting the atributtes and filters of interest. The available filters are:

In [20]:
from pybiomart import Server

# Connect to the server
server = Server(host='http://www.ensembl.org')

# Select the dataset
dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['hsapiens_gene_ensembl']

# List all available filters
filters = dataset.list_filters()
filters


name                    type  \
0                         link_so_mini_closure                    list   
1                              link_go_closure                    text   
2            link_ensembl_transcript_stable_id                    text   
3                                      gene_id                    text   
4                                transcript_id                    text   
5                         link_ensembl_gene_id                    text   
6                              chromosome_name                    text   
7                                        start                    text   
8                                          end                    text   
9                                   band_start  drop_down_basic_filter   
10                                    band_end  drop_down_basic_filter   
11                                marker_start  drop_down_basic_filter   
12                                  marker_end                           
13                        hsapiens_encode.type                    text   
14                                      strand                    text   
15                          chromosomal_region                    text   
16                       id_list_xrefs_filters            boolean_list   
17                 id_list_limit_xrefs_filters                 id_list   
18                  id_list_microarray_filters            boolean_list   
19            id_list_limit_microarray_filters                 id_list   
20               transcript_count_greater_than                    text   
21                  transcript_count_less_than                    text   
22                                     biotype                    list   
23                          transcript_biotype                    list   
24                                      source                    list   
25                           transcript_source                    list   
26                              transcript_tsl                 boolean   
27                    transcript_gencode_basic                 boolean   
28                           transcript_appris                 boolean   
29                     transcript_is_canonical                 boolean   
30                                 mane_select                 boolean   
31                          mane_plus_clinical                 boolean   
32                       phenotype_description                    list   
33                            phenotype_source                    list   
34                                     go_clos                           
35                                     go_name                           
36                            go_evidence_code                    list   
37                             homolog_filters            boolean_list   
38  id_list_protein_domain_and_feature_filters            boolean_list   
39        id_list_limit_protein_domain_filters                 id_list   
40                  germ_line_variation_source                    list   
41                    somatic_variation_source                    list   
42                          with_validated_snp                 boolean   
43                         so_consequence_name                    list   

   description  
0               
1               
2               
3               
4               
5               
6               
7               
8               
9               
10              
11              
12              
13              
14              
15              
16              
17              
18              
19              
20              
21              
22              
23              
24              
25              
26              
27              
28              
29              
30              
31              
32              
33              
34              
35              
36              
37              
38              
39              
40            

If we want to get information on a gene by 'gene_id' we can use:

In [23]:
# Define the gene id
gene_id = 'ENSG00000001626'

# Define the attributes
attributes = ['ensembl_gene_id', 'external_gene_name', 'description', 'chromosome_name', 'start_position', 'end_position', 'percentage_gene_gc_content', 'gene_biotype']

# Query the dataset
result = dataset.query(attributes=attributes, filters={'gene_id': [gene_id]})

# Print the result
result


Empty DataFrame
Columns: [Gene stable ID, Gene name, Gene description, Chromosome/scaffold name, Gene start (bp), Gene end (bp), Gene % GC content, Gene type]
Index: []

In [26]:
from pybiomart import Server

# Connect to the server
server = Server(host='http://www.ensembl.org')

# Select the dataset
dataset = server.marts['ENSEMBL_MART_ENSEMBL'].datasets['hsapiens_gene_ensembl']

# Define the gene id
gene_id = 'ENSG00000139618'  # replace with your gene id

# Define the attributes
attributes = ['ensembl_gene_id', 'external_gene_name', 'description', 'chromosome_name', 'start_position', 'end_position', 'percentage_gene_gc_content', 'gene_biotype']

# Query the dataset
result = dataset.query(attributes=attributes, filters={'gene_id': [gene_id]})

# Print the result
print(result)


Empty DataFrame
Columns: [Gene stable ID, Gene name, Gene description, Chromosome/scaffold name, Gene start (bp), Gene end (bp), Gene % GC content, Gene type]
Index: []


Not easy to read... Don't worry you can convert the 'genes' object into a pandas DataFrame easier reading and manipulation. Here’s how you can do it:

In [ ]:
import pandas as pd
# Convert the 'genes' object into a pandas DataFrame.
genes_df = pd.DataFrame(genes)

# Display the DataFrame.
genes_df


## Exercise 1: Exploring Gene Information

Now that we have our data in a more readable format, let's start exploring it. For this exercise, you will retrieve and analyze information about a specific gene.

**Task:**

1. Retrieve information about the gene with the Ensembl ID 'ENSG00000139618'.
2. Display the information in a readable format.
3. Analyze the information and answer the following questions:
    - What is the name of the gene?
    - On which chromosome is the gene located?
    - What is the start and end position of the gene on the chromosome?
    - What is the percentage of GC content in the gene?
    - What is the biotype of the gene?
    - What is the status of the gene in the BioMart database?

Write your code in the cell below:

In [ ]:
# Write your code here


## Exercise 2: Comparing Genes

In this exercise, you will compare two genes based on their GC content and biotype.

**Task:**

1. Retrieve information about the genes with the Ensembl IDs 'ENSG00000139618' and 'ENSG00000248333'.
2. Compare the two genes based on their GC content and biotype.
3. Answer the following questions:
    - Which gene has a higher GC content?
    - Do the two genes have the same biotype?

Write your code in the cell below:

In [ ]:
# Write your code here


## Scenario-Based Exercise: Investigating a Genetic Disorder

Imagine you are a geneticist investigating a rare genetic disorder. You suspect that the disorder is caused by a mutation in one of two genes: 'ENSG00000139618' or 'ENSG00000248333'.

You decide to use the BioMart database to gather more information about these two genes.

**Task:**

1. Retrieve information about the genes 'ENSG00000139618' and 'ENSG00000248333'.
2. Analyze the information and answer the following questions:
    - On which chromosomes are the genes located?
    - What are the start and end positions of the genes on their respective chromosomes?
    - What are the bitypes of the genes?
    - What are the statuses of the genes in the BioMart database?
3. Based on your analysis, do you think one of these genes could be responsible for the genetic disorder? Why or why not?

Write your code in the cell below:

In [ ]:
# Write your code here


## Conclusion

Congratulations on completing the tutorial! 🎉 You've learned how to access and use the BioMart database using Python. You've also gained experience in analyzing gene information, which is a crucial skill in bioinformatics and pharmacogenomics.

Keep practicing and exploring the BioMart database. There's a wealth of information waiting to be discovered! 🚀
